// jsmm-applet/lib/index.js
/*jshint node:true jquery:true*/
"use strict";

module.exports.clayer = require('./clayer');
module.exports.dares = require('./dares');
module.exports.editor = require('./editor');
module.exports.info = require('./info');
module.exports.jsmm = require('./jsmm');
module.exports.output = require('./output');
module.exports.robot = require('./robot');

module.exports.UI = function() { return this.init.apply(this, arguments); };
module.exports.UI.prototype = {
	icons: {dare: 'icon-file', console: 'icon-list-alt', canvas: 'icon-picture', robot: 'icon-th', info: 'icon-info-sign', home: 'icon-home'},
	paneOutputs: ['robot', 'console', 'canvas', 'info'],
	constructors: {
		robot: module.exports.output.Robot,
		console: module.exports.output.Console,
		canvas: module.exports.output.Canvas,
		info: module.exports.info.Info,
		input: module.exports.output.Input,
		Math: module.exports.output.Math
	},

	init: function($main, options) {
		this.$main = $main;
		this.$main.addClass('ui-main');

		this.options = options;

		this.$output = $('<div class="ui-output tabbable"></div>');
		this.$main.append(this.$output);
		this.$tabs = $('<ul class="nav nav-tabs"></ul>');
		this.$output.append(this.$tabs);
		this.$content = $('<div class="tab-content">');
		this.$output.append(this.$content);

		this.$editor = $('<div class="ui-editor"></div>');
		this.$toolbar = $('<div class="ui-toolbar"></div>');
		this.$main.append(this.$editor);
		this.$main.append(this.$toolbar);

		this.outputs = [];
	},

	remove: function() {
		this.removeAll();
		this.$main.removeClass('ui-main');
		this.$output.remove();
		this.$editor.remove();
		this.$toolbar.remove();
	},

	removeAll: function() {
		for (var i=0; i<this.outputs.length; i++) {
			this.outputs[i].remove();
		}
		this.$tabs.children('li').remove();
		this.$content.children('div').remove();

		this.scope = {};
		this.outputs = [];
		this.outputsByName = {};
		this.tabs = [];
		this.tabsByName = {};
	},

	load: function(options) {
		this.removeAll();
		this.addEditor(options.editor);

		for (var name in options.outputs) {
			if (name !== 'dare') {
				var output;
				if (this.paneOutputs.indexOf(name) >= 0) {
					output = new this.constructors[name](this.editor, options.outputs[name], this.addTab(name));
				} else {
					output = new this.constructors[name](this.editor, options.outputs[name]);
				}
				this.outputs.push(output);
				this.outputsByName[name] = output;

				if (name === 'input') {
					this.scope.document = output.getAugmentedDocumentObject();
					this.scope.window = output.getAugmentedWindowObject();

					var mouseObjects = options.outputs[name].mouseObjects || [];
					for (var i=0; i<mouseObjects.length; i++) {
						var outputName = mouseObjects[i];
						output.addMouseEvents(this.outputsByName[outputName].getMouseElement(), outputName, this.scope[outputName]);
					}
				} else if (output.getAugmentedObject !== undefined) {
					this.scope[name] = output.getAugmentedObject();
				}
			}
		}

		this.$tabs.toggle(!options.hideTabs);

		this.editor.updateSettings(new module.exports.jsmm.Runner(this.editor, this.scope, {}), this.outputs);
		this.selectTab(this.tabs[0]);
	},

	addTab: function(name) {
		var $tab = $('<li></li>');
		setTimeout(function() { $tab.addClass('tab-button-enabled'); }, 200*this.tabs.length);
		this.$tabs.append($tab);

		var $link = $('<a href="#"><i class="' + this.icons[name] + ' icon-white"></i> ' + name + '</a>');
		$tab.append($link);

		$link.click($.proxy(function(event) {
			event.preventDefault();
			this.selectTab(name);
		}, this));

		var $pane = $('<div class="tab-pane"></div>');
		this.$content.append($pane);

		var $output = $('<div class="tab-output"></div>');
		$pane.append($output);

		this.tabs.push(name);
		this.tabsByName[name] = {$pane: $pane, $tab: $tab};
		return $output;
	},

	addEditor: function(options) {
		this.editor = new module.exports.editor.Editor(options, module.exports.jsmm, this.$editor, this.$toolbar);
		this.$toolbar.removeClass('ui-toolbar-enabled');
		var $toolbar = this.$toolbar;
		setTimeout(function() { $toolbar.addClass('ui-toolbar-enabled'); }, 200);
		return this.editor;
	},

	/*
	addDare: function(dare) {
		var $pane = this.addTab('dare');
		this.dare = dare;
		this.outputs.push(this.dare);
		dare.makeActive($pane, this);
		this.$main.addClass('ui-dares-active');
		return this.dare;
	},
	*/

	selectTab: function(name) {
		this.$content.children('.active').removeClass('active');
		this.$tabs.children('ul li.active').removeClass('active');
		this.tabsByName[name].$pane.addClass('active');
		this.tabsByName[name].$tab.addClass('active');
		if (this.outputsByName[name].setFocus !== undefined) this.outputsByName[name].setFocus();
	},

	loadDefault: function() {
		this.load({
			editor: {},
			outputs: {
				robot: {},
				console: {},
				canvas: {},
				info: {},
				input: {mouseObjects: ['canvas']},
				Math: {}
			}
		});
	}

	/*
	loadInitial: function() {
		this.removeAll();
		this.addEditor();
		this.addRobot();
		this.addConsole();
		this.addCanvas();
		this.addInfo();
		this.addInput(['canvas']);
		this.addMath();
		this.finish();

		this.editor.setText(window.localStorage.getItem('initial-code') || '');
		this.editor.setTextChangeCallback(function(text) {
			window.localStorage.setItem('initial-code', text);
		});
		if (window.localStorage.getItem('initial-robot') !== null) {
			this.robot.setState(window.localStorage.getItem('initial-robot'));
		}
		this.robot.setStateChangedCallback(function(state) {
			window.localStorage.setItem('initial-robot', state);
		});
	},

	hideDares: function() {
		this.dares.hide();
	},

	getCanvas: function() {
		return this.canvas;
	}
	*/
};
