// jsmm-applet/lib/dares/dares.dares.js
/*jshint node:true jquery:true*/
"use strict";

module.exports = function(dares) {
	dares.Dares = function() { return this.init.apply(this, arguments); };

	dares.Dares.prototype = {
		init: function($collection, content) {
			this.content = content;
			this.$collection = $collection;
			this.$collection.addClass('dares-collection');

			//this.$header = $('<div class="modal-header"><a href="#" class="close" data-dismiss="modal">&times;</a><h3>Select a Dare</h3></div>');
			//this.$collection.append(this.$header);
			this.$header = $('<div class="dares-header"></div>');
			this.$collection.append(this.$header);

			var $title = $('<div class="dares-header-title">' + content.title + '</div>');
			this.$header.append($title);

			var $difficulty = $('<div class="dares-header-difficulty"></div>');
			for (var i=0; i<content.difficulty; i++) {
				$difficulty.append('<i class="icon-star-yellow"></i>');
			}
			this.$header.append($difficulty);

			this.$body = $('<div class="dares-body"></div>');
			this.$collection.append(this.$body);

			for (i=0; i<content.dares.length; i++) {
				var dare = content.dares[i];

				var $item = $('<div class="dares-body-item"></div>');
				if (dare.completed) {
					$item.addClass('dares-body-completed');
				}

				//$item.data('dare', dare);
				$item.on('click', $.proxy(this.itemClick, this));

				var $name = $('<span class="dares-body-name">' + dare.name + ' </span>');
				for (var j=0; j<dare.outputs.length; j++) {
					var output = dare.outputs[j];
					$name.append('<span class="dares-body-output"><i class="' + this.icons[output] + ' icon-white"></i> ' + output + '</span>');
				}
				$item.append($name);
				$item.append('<span class="dares-body-highscore"><i class="icon-trophy"></i> ' + dare.highscore +'</span>');

				this.$body.append($item);
			}
		},

		remove: function() {
			this.$header.remove();
			this.$body.remove();
			this.$collection.removeClass('dares-collection');
		},

		icons: {console: 'icon-list-alt', canvas: 'icon-picture', robot: 'icon-th'},

		itemClick: function(event) {
			var $target = $(event.delegateTarget);
			var dare = $target.data('dare');
			var $preview = $target.children('.dares-body-preview');
			if ($target.hasClass('dares-body-item-active')) {
				$target.removeClass('dares-body-item-active');
				$preview.slideUp(200);
			} else {
				this.$body.find('.dares-body-item-active').removeClass('dares-body-item-active').children('.dares-body-preview').slideUp(200);
				$preview.show();
				dare.setPreview($preview);
				$target.addClass('dares-body-item-active');
				$preview.hide();
				$preview.slideDown(200);
			}
		},

		selectDare: function(body, number) {
			this.list.bodys[body].dares[number].selectDare();
		}
	};
};
