// jsmm-applet/lib/dares/dares.dares.js
/*jshint node:true jquery:true*/
"use strict";

module.exports = function(dares) {
	dares.Dares = function() { return this.init.apply(this, arguments); };

	dares.Dares.prototype = {
		init: function($collection, content) {
			this.content = content;
			this.$collection = $collection;
			this.$collection.addClass('dares-collection');

			//this.$header = $('<div class="modal-header"><a href="#" class="close" data-dismiss="modal">&times;</a><h3>Select a Dare</h3></div>');
			//this.$collection.append(this.$header);
			this.$header = $('<div class="dares-header"></div>');
			this.$collection.append(this.$header);

			var $title = $('<div class="dares-header-title">' + content.title + '</div>');
			this.$header.append($title);

			this.$body = $('<div class="dares-body"></div>');
			this.$collection.append(this.$body);

			for (var i=0; i<content.dares.length; i++) {
				var dare = content.dares[i];

				var $item = $('<div class="dares-body-item"></div>');
				var $cell = $('<div class="dares-body-cell"></div>');
				$item.append($cell);

				var $preview = $('<div class="dares-body-preview"></div>');
				$preview.hide();
				$item.append($preview);

				//$item.data('dare', dare);
				$item.on('click', $.proxy(this.itemClick, this));

				var $arrow = $('<span class="dares-body-cell-arrow"></span>');
				$cell.append($arrow);

				var $name = $('<span class="dares-body-cell-name">' + dare.name + ' </span>');
				for (var j=0; j<dare.outputs.length; j++) {
					var output = dare.outputs[j];
					$name.append('<span class="dares-body-output"><i class="' + this.icons[output] + ' icon-white"></i> ' + output + '</span>');
				}
				$cell.append($name);

				var $difficulty = $('<span class="dares-body-cell-difficulty"></span>');
				for (j=0; j<5; j++) {
					$difficulty.append('<i class="icon-star' + (dare.difficulty <= j ? '-empty' : '') + ' icon-white"></i>');
				}
				$cell.append($difficulty);

				//$cell.append('<span class="dares-body-cell-completed"><i class="icon-user icon-white"></i> ' + dare.completed +'</span>');
				$cell.append('<span class="dares-body-cell-highscore"><i class="icon-trophy icon-white"></i> ' + dare.highscore +'</span>');
				$cell.append('<span class="dares-body-cell-completed">' + (dare.completed ? '<i class="icon-ok icon-white"></i>' : '') + '</span>');

				this.$body.append($item);
			}
		},

		remove: function() {
			this.$header.remove();
			this.$body.remove();
			this.$collection.removeClass('dares-collection');
		},

		icons: {console: 'icon-list-alt', canvas: 'icon-picture', robot: 'icon-th'},

		itemClick: function(event) {
			var $target = $(event.delegateTarget);
			var dare = $target.data('dare');
			var $preview = $target.children('.dares-body-preview');
			if ($target.hasClass('dares-body-item-active')) {
				$target.removeClass('dares-body-item-active');
				$preview.slideUp(200);
			} else {
				this.$body.find('.dares-body-item-active').removeClass('dares-body-item-active').children('.dares-body-preview').slideUp(200);
				$preview.show();
				dare.setPreview($preview);
				$target.addClass('dares-body-item-active');
				$preview.hide();
				$preview.slideDown(200);
			}
		},

		selectDare: function(body, number) {
			this.list.bodys[body].dares[number].selectDare();
		}
	};
};
